src/GridTurbulence.cpp

```diff
@@ -20,7 +20,7 @@ double turbulentCorrelationLength(double lMin, double lMax, double alpha) {
 
 #ifdef CRPROPA_HAVE_FFTW3F
 
```

Comment:
Probably we should have two functions then for both vector grids.

Reply:
This is what bothers me: should we do code duplication in these cases or use templates? If we use templates, however, we have to move a lot of code inside of the header files, which is messy. The 3rd option is that we keep only the double version of functions (if the float one is not explicitly necessary), and rely on the C++ automatic casting of float to double, right?

Reply:
The most straight forward way would be to avoid code duplication and use templates for the functions.
To avoid clutter in the header files you could separate the template implementation into an additional subfolder (e.g. details) using forward declaration:
`// foo.h
template <typename T> foo(T f);

#include detail/foo.cpp;
`
`
// detail/foo.cpp
template <typename T> foo(T f)
{
// ...
}

 This should probably also be done for the functions in gridtools.

Reply:
Sorry for the late reply.
I think keeping the simple ScalarGrid and VectorGrid in future versions makes everybody's life easier and the naming convention cleaner. Can't we just call ScalarGridf and VectorGridf for that?
For most applications float precisions is enough, anyways.
This is just an extra line in SWIG.

Reply:
Just to be clear, this comment is a reply to Tobias' comment:  "deprecation warning that VectorGrid will be removed in CRPropa 3.9 or so nd users should use VectorGridf / ScalarGridf instead."

Reply:
The API should be easy to understand and there should be thus consistent names. To the user, and also future developers,  it is totally opaque why there is inconsistent naming. Sure, this one thing is not a big issue, but many small issues will make a mess at some point. Users already have to use Vector3d and Vector3f and thus think about the precision they want, so I don't think that we ask too much of the users here.

Reply:
I agree that it this is not a lot to ask. But ScalarGridf is aesthetically very unpleasant. It seems like a typo. For Vector3f, the `f' appears after a number, which makes it look okay. If we just start appending f's and d's after objects, things will get ugly. 

I think one of the nice things about CRPropa is that most of the code looks nice and clean. 
If we are asking the users to adjust to a new notation anyways, can't we think of something nicer? 

Grid1f for ScalarGridf and Grid3f for VectorGridf? This would also make the implementation of grid for variable photon fields more natural: GridNf/GridNd (N is not a number is capitalised, so it is visually better).

Reply:
I agree with both of you in your arguments: VectorGridd really looks like a typo, but we should strive for consistency across the code. @rafaelab's suggestion meets both requirements, thus if there is no other suggestion that goes in the same direction, I would vote for Grid(dim)(d|f).

Reply:
Grid(dim)(d|f) sounds great.